{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/kkokay07/GenomicClass_on_Cloud/blob/master/NGS_Variant_calling.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Creator: Dr. KANAKA K. K.\n",
        "![Kanaka-photo-1-229x300.jpg](data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEAYABgAAD//gA7Q1JFQVRPUjogZ2QtanBlZyB2MS4wICh1c2luZyBJSkcgSlBFRyB2NjIpLCBxdWFsaXR5ID0gODIK/9sAQwAGBAQFBAQGBQUFBgYGBwkOCQkICAkSDQ0KDhUSFhYVEhQUFxohHBcYHxkUFB0nHR8iIyUlJRYcKSwoJCshJCUk/9sAQwEGBgYJCAkRCQkRJBgUGCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQk/8AAEQgBLADlAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/EAB8BAAMBAQEBAQEBAQEAAAAAAAABAgMEBQYHCAkKC//EALURAAIBAgQEAwQHBQQEAAECdwABAgMRBAUhMQYSQVEHYXETIjKBCBRCkaGxwQkjM1LwFWJy0QoWJDThJfEXGBkaJicoKSo1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoKDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uLj5OXm5+jp6vLz9PX29/j5+v/aAAwDAQACEQMRAD8A+qaKKKACiiigAooooAKKKKACiiigAoprSIilmYKo5JPavPfF/wAZdF8PlrfTyNTvAcYjb92p92/woGk2eiUma+atZ+Nvi27kZ4b63sIu0cUYJ/M81lQ/F3xeW3DXZyfcLj8sUFcjPquivl9fi/4wU/Pq8n/fC/4Vp2Xxw8V2xBmnhnT0khXn8RQHIz6NorxrRf2gkklSLV9LZFPBmt2z+O0/416fofijSPEUAl02+inB6qDhh9R1oJaa3NaiiigQUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRSMwVST0FAAW2gkkACvLvG3x00vw/PLYaRANTvU+UsDiJD7kdfwrlvi18Z7mN7rQNKtngT7ktw7FXYdwB2B9c141HqbOp2wqrZ6pQaRj3Ov8RfEXxL4q3R6pc7LZjnyIvlRfbGefxzXNNcKMqN2PU8VVaUSN80ilveSo2nSEkHY305osWtNi0SjqQSefQ/8A1qaqRR8l3JPYn/61QR3isSAgHfGKiN7bs+SdrZ9KBlt7tIl7ewPNIuoyrx9nR1P90VA99bocSA5HXPapIJLacl0nBP8AdNILlpUS4Xcm5D6DpViwv9R0u5S5srmSCZD8ro2DVQyGL5gMqOTt649RUm8bd+RJH1J9KCtGe2eB/jvzFYeKItrHCi8jHH1cf1FeyWl5b30CXFrNHNC4yro2QRXxaWbarwkFW7E133ws+IN3omqRaY9w6Wk77dkhJVWPHHtmnuZSp9UfTQoqnpWoJqNlHcLgFuoB6HvVygyCiiigAooooAKKKKACiiigAopM0tABRRRQAUUUUAFc5468Uw+EvD9xfybWlI2Qxk/fc9B9O5roia+ZPjV4rfXfGdxpyz4tLBTCiqeC3Vj9c8fhQVFXZ5t4g8Qzaxqlxd3DGWaRyzu3GT7D0qgt+SuwKEB74qldbTK38PPSo4i6tnOfamjVlmUOJtoOM/xetWrWMhvnKmmfJPECqncPfmpUC7NxXPY5GOaBMR7kwMeA/OcY5FLHq1ncsI32xv23dD+NMmt0n+UMR7g5xWfNozGThw59DxSbGkzekjtp1EUka46rk459jWfNaLaP+7IwTnBOP1psVpOqCNyVC8c/MKna3dAN/wA49jz+FZuotmWqbepB51xGfNtpSV6NGecVdtLwhdxUhW7Z4qvttzlxujYenQ06KMOcA8Dg47+9HONQuSm52KYAcI7ZB9DUtjqX2e7jaU/dYENVWexmkjyqkqfSqyvsHlzDGe56g0RmmOUGtz7H+H/iSy1fSolt5A7KozyM9OTjtXYgg18yfArV7dtaS0d1jdx8hcAFiOwOODivpmLIAyc+9bM5ZIfRRRUkhRRRQAUUUmaAFyKKaaKAFozRRQAuc0UlFAC0fjSUtADWOB0zXw/4hklk1rUp5iWkNxIfx3GvuBuQa+JNftGTX9Rt5cgx3MgP13Gg0gYMVql3JtOSx79hXR6b4AvbtRJGwAPQGpfCenRahqe4oBHD29TXq1nEkSAKoAFedi8TKEuWJ6eFw0Zx5pHn9r8OLxmAZQfUg1qw/CGWcbnuBGe2K7+224HrWzZqGGDgVw/Wqre52/VqSWx5VL8HrhRuS5BYDvWfL8KtTVgco4Hr2r3Ixpt55qLaQegxWixFXuR7Cn2PFG+H13br+9Bf2Wq58J3kj7Y4OnHINe5PbxSAZQflUbWcQ48tQKPbTe41Sgjxf/hX87LmWI7j1AHFWLX4eogy3UenSvXJbaNgQV4qhdIig7Qo+gqJVp9zRUo9jh4fCltDENqDcOq9q4XxtoIsp0lRMBuDgV7DLtVSK5LxtY+dpzSgZ2/MBWtCbUkzKtBOLR594eu59Lv7e4iYh433DBr7N8OXx1LQ7K8bOZoVfnryK+LxKizxmMZJfaRX2b4Vha38OaZFInlslrGCvodor2VseHURq0UUUGQUUmaKACiiigAooooASiiimMWiiikISnUlGaAA9K+OfHyLD4312Mc/6bLz9WNfY2M18k/HWxbSfH+pxouBeMlwpx2YDP6g0FwE8BW2xLiQgbmkIH0BrvoW+TiuF8DBjbheSVGK7hGCDrXi4le+z38NpBF+3bkVsWcgBrAimzjbWjbvK2No61go2Zs3c3xLwPlpnm5bpVJFuNuDmmsZYjnFaqxnZmnubHSoZHw3PeoY7pyAGFI4eQ5Ap6BYZPPsQ5rFurnOT0rUuoJGXhTWHeRSL1U1nyplp2IXkLHrWb4iZTpcqt3WrDO6tgg1Br8Bm0aZx1Vc1vTjZoxqPQ868FaT/a/jSw04kbZryPOfTcCf0r7PQBVAAwBwK+WvgRZrefEmF5QG8mOSVc+oGB/OvqbFeutjwqr1FpDRSUzMKKWkoAWkoooAKKWigYh5oApKXNMApaSigBaKSikICwUEk4A5zXz3+0NNo2p3mmXlm3mXiFopHCnaUHI578k17vrNwLTSrqY/wxNgepxxXhPinwv/AGh4YMYJ3wK0kbfT+lc1bEKnJRfU7cLhnUjKfY5zwSgEVy6jgMAK35ZSW2g4GawfAisdMuXb/npitHUoXkTCuyg9cVxV7c7PSw91BG3a6jptrgTXKFu4B6V0el63o0hAFzCMdcsK8fnFhYPta1kupnP3UByfrWZ4ge6025FrcWDaY7IrqZbo4IYgDgBvXP0BpQoxntcVSrKB9EJq2kSN5cV5buR2DA0ya5tn4SSNh04INfO1j4Y8Q32jQ6yp8q3kkaJXRu4Pf65610um23iS2CoZeOxY4NZ1qSg9GVQnKau0euGSAHJIFWFvbKNMySIoHUk4xXmxk1xV8tpIwcZ3F65/U7XxBcxzxtNujI+YqeMVkk27M3ltoey3Gv6HbRky31sOP745rntU8XaGFZllSQf7JBrxDX/Buu+HoLO+vWMVvdo7xFyxJI6DA6Z/+vWx4f0rVJ9LutTtrC0vbS3fYHdmVpPdQ3fHb3rr+rQUea5wqtPm5Tt28Q6RfvsimEb54Dd6muR5ukXqDkeSxH5VyNjLDcjL2H2dv7u3vXYaWoltJI9p5jZefpRCyNJ3tqZPwFvLPTPFF7fXwKIsXlJLj5VLNzn8BX0urBgCDx1r5z8H6HHH4fM7gfvpCx/A4/pX0LYSCayglAIDxqcHtxXbSr88pR7HnYmhyRjPuWKKSiug4xaKSigAooozTGFFJRQAUUUUxgKWkopALRRRQIralZrqFhPascCVCufT3ry6+drWymgdQHVGhIbpnnNetV5f46gktNUuBGm7eRMq+oPXH4g152YU7xU1uj08sqWm4PZnnfhO28qwnj24/enNdCunxyRbXQNVDSypkutqFAZM7T1B710+mRLIATg/WuSs9mehTSV0Yn9k+Su6C2Tg56daq6vp9nrSxDUNJluJIhtQ+WflHoCD0rtZbAnmNsVVe3vnzGpRV9duTXPGtZmzp3RiW6TNYRafHHJa2MQwsKyEDHptB/U1WulkYpGn7tE+UZOSB6V039mm2gaSaU5xyTWMqJLMgznJpVakpFUqaWwyS1/0fcU5A64qCxtkmV0kjEqHkgjOD64rpvshki27eCKwJR/Zt5t5GTnj0rKEpXNGk0Ov4GuLcQSWFveW6nO0AHn3ByM1RcTpB9mgsWtox0URhQPy4rq00pL2BZ4JGDEfeU4qrJpt8SUklDL645ro9s9pHP7NX0Oai04sw3xDI7kVpWsCxttC4GK1lsBCvJy3eqtwoiLtwMKT+laU53krEVI2RW8FWxmsbfTyv3pdoH1bOa9oiQRRqi8BQAK8v+HFnJLqELyYOxPNOBwOOP516lXo4ONouT6s8rMJ3koLZIKKWkrtRwIKKKKACkoooGFFFFACUtNFKKYxaKKKQhRRSA0tAgri/iJAgFncgHzMsn1HWu0rM17RU1q1ERk8uRDuR8ZwfcVjWi5QaRtQmoVFJ7HhCXuNTuwePnFdJpty/wC72dM8/SsbxfocnhrX3jllSUzKsgZFwM9MVPpd4DxmvNrw5Yq57NGak7o7yxdX+9irUnlxKW4+tYtjc7Vznj0qxd3qLAXkYLGBlia4E0dEoa3MTW76aUeYwb7OD8wUc4rMtvEfh+afbFc/vE6qylRn6kVm6/8AELT4Xa3tpc7fQda87uvH93PeZiSNV3YPyDpXRHDSmS8VGCsj6Jtb6xa0UkjOPWuJ1jxho8WpmCQSO33V2Rlh+dcN/wAJKH0ie7gnlQQkIYg3y5P8utc1H431G0vGKlGRuR8vT0xVRwsnuZ/WoR1XU930K+ltyHCsIXOQhGCB9K6jMU0e8AV43oPxOtmMUd5G4dvl3cYH1r0u21FZIgVcEEdqyqQlDdGikpv3R1+wTJzXP6ldAW1wc/wN/Kr+qXg2nng1z426hcx2e44mZUOPQnFaYVc0icRK0dTufhdI4DRSr8zQJhvYf/rr0OsrQvDlloEHl2odjgAvIcsQK1a9mlBxjZng15qc+ZBRRRWpkFJSmkoGFFFIaYwooxRQA2lHWigUxjhRSClpEgKWk70tIAooopCZ5P8AHO1a1jsNURcqS0L8dD1Xn86890rUfLHzENnoc9BXunxG8PP4m8JX1jFzOF82H3deQPx5H4180aTqUaGRJXKSKSrKeoIrkxMLq6PQwdS2jPV9KvRIgAbPFc38SvEE1paJZwH5pMhsdhil0HU1yqjIT371V8YwR31zaKjqZJG27RyfrXl0qdqmp6lSV6dkcLplnbzTE3PmFn5AAyT9a6rSNA05EAfSmkU9/KZ/zrrINKs7W0QZUsi8SYwalg1C0s8B5IdvGS3HFdTrczIp0YxWpk/2VpkcboNBhDeghbB+orEuNCsISzLpcjgfMS0Xyj2+ldu/izQdxHnw7un3jVC712zumZbdodpGM9aNYl2ps8u1qxjmkaWCxliCfewpGB613nw91O4uLHypHaQIOp9KdqKR/wBnXAjKnCksc8k1U8F3qW9pOBjJY8DjjtRUlz02jGMeSombeu6gUyuecUzwIjaj4u021B3gN5sg9Aozn88VzetapidyWz357V6P8CNDaVb3xHOnM3+jwH/ZGCxH44H4Vrg6VlcwxtXRo9eooor0DyAooopgFJS0lAwpKWkpjEooopgFApBSigGKKWkpaQgpaSlpCCiiuY+I3ipfCXhe5vEcC6kHk249XPf8Bk/hTUXJ2Qm7HO/EH436J4JmexiibUL1eGRG2pGfQn19hXzTrniE6/rd7rdtbR2vmy75IYQdi57jP0rF8Z3cs0gld2dmclmJySfWrPw7njuNRubWYB0ngIKnvgita8I04tFULuSsdfo/iRGg3SThWTj60qeJxPfCXdmRTsXJ6VyOvabL4fv+BvgPKZzgj8O9V7e9BdXMuzGM8ZPTtXnxoQfvI75V5L3ZHt1vryX2niAt++OAR2FcJ4tlvYZtsZkZQMsF4yKo6JrgshwGZzghye2avxaxLqzFJnRAclnHOPYZrJUeSXMbe3548tzm43vDEsz+YoLhVP511Hhy01FgJJ88MB06D1pby809II1jUFy+3Gcn6/rTxrsFragQtjeMFmb2xx+tazbkrJGUIqErtm1rUgs7EojnzG6kHnFcpp+rLp3nSBgoVtoC/Ss/VdflaQhm3nbwQ3APasF72SdHjTneccevHP6UU6Fo2YquIvK6OhvNUk1a4MduS2c7sdgOf6V6d8KPjSdHhg0fVgH09fljlUfNCPf1FcJoGhppmkzzyL/pE0TZz/CMGuG0K7fYM9VODXdg1CbcDkxSkkpM++bW6hvLeO4t5FlikUMjqchge9S14b8BvHTmY+HL2UsjjfbFj91u6/j1/CvcqKtN05cphCV0FFFFZlAaSiigYU00ppDTGhKKM0UxhS03NL2oBjs0tNFL0pCFozSZ4rlvFPxL8OeE42+2XqS3AHFvAQ8hPuBwPxxQot6IltI6rPvXz98etfN94ki0lHJhsIhuAPHmPyf021Q8VftC6/f74tFt4dKgOQJGxJL9cngfl+NefSalc6tcS3V5cyXNxKd0kshyzn1NehhcNJT5pGFSppZHOeI7Q3Fq+3qORVrwHYpaKLvO6WQlSfQelX7u2EqFcZzVHw7IbK+ks24BO9KyzKk1HmRvgZrmszub/S4dZtTBOOCPlbHKn1rgda8PXuguVOx4ScpKB3/pXo1hJ5sYA61ea3juIWhnjDowwQa8CFZ03boe1PDxqq/U8VgkuIVL/eXGMZ6fjV2DU7yyh3rjY4IAyD1712Ws/Du2uCzWVwbbP8DKSp/KuSvvBes2nyRwGUf3oj1Fdsa8JI4JYacSD+1l80jgPJwD2X/OKWTU5JWkjIGCeB2HPvTIfCWtNkCxkUnjc3GK29N+H19cHddzCIE/NgZJ/Gq9rBa3JVGo+hz3mlsKpLSM2cdcGu08K+FUjSO+uhlsApGf5mtvSfBOn6cA+ze4GNz/ANK13CR4RFrjrYtP3YndQwdtZmZq0oitnA7jFeWaaB9onx08xv516L4hl8m0mmc8KpOa4PRbYlN7Dljk135VHmk2cmZOyR0mkX1xp0sd1azPDNEQyOhwVI7ivRfCX7SOoaVP9h8SIl/Ep2+cmElX69j+lecRx7V6VyWrOYNWZh91sZr2q8ItLmR5FNu59peH/i34R8RBFt9VSCV+kdyPLOfTJ4P512CurqGVgynkEHg18HWswQZBwjV3ng34s+IvCQ8i3vBc2o/5d5zuVfp3H4Vxywml4s1VTufW1FeJ6X+0bGSv9paODGfvSW8vI/4Cf8a9H8M/ETw54tAXTdRjM5628vySD/gJ6/hXPKjOO6NVNM6SkNGaTNZliUUZopjClzxSE1g+KvG2jeDrM3Gp3QVyP3cCcySfQenueKEm9EDdjfJAGSQBXA+L/jN4d8Nb4LeQ6lerx5UBGxT/ALTdB+GTXivj342ar4peS1t5GtLDoLa3blx/tv3+nSvOpLy4mPURL6Lyfzrrp4XrMwlUfQ9E8XfGvxDre9HvPsVu3S3tcrke5+8fzxXnc+q3d25KLjP8Uhyag2AEnuepPU0NJxtXiuyMVHbQxbb3Kt2Xc7pJGcj8APwrY0q5yBzWRcLmM07TLgocelVB2YpK6Ot4YZFZmoWjrKlzDxJGcirNtch0HNWOHGK3qQVSLizKEnCV0bWkXi3Nuk8Zwf4h6Guv0547hQGHNedadMdLud+CYmPzCuxsdXsiVaO5jHAPzHFfH43B1aUmkro+owuKp1Ip3sdOdG8wZTp7f/XqM6HNG3MaH65FW9I1VJAAsiSL/ssDW7HdKw+6R9K8mUnHRnerPY5OTS3xxEmfrTU0i4PSMD6DpXXvIrD7ucVFJMOy1HtCuU5iTRJFGZHwPbk1UuLNUXCrjFdHdBpM8YFY+qSw2Fq8s7hcDvVwk5O0ROyV5HnXjk7bVbRfvzNz9KxLK1ESKMYxV++uTq9+92/KA7UHoKQ7VGMivtsswjo0ry3Z8rmGIVWp7uyI5G2pXGao3nX7fWusvZdkLEelchFm4vHbrzXVW10OWmXbRj5e09qsFY5sbhz2PpUAGx8U5TtfB6GklYon8+WDgMePXmrFtqdxDIskbFHU5Dxkgj6VTkBKZ64qIMQeKQHtXgX4/ajpHl2mvh9QtBwJf+WyD69G/GvcvD3jfw/4niEml6nbzMRkxltrr9VPNfFkcueDUsN49vIGjZkYcgqcEVz1MPGWuxpGbR91cetFfI+i+PvEMNkFi1i9VQSAPOb/ABoqVgZPqJ4jyPYfiT8brLw55ml6D5d9qXRpesUB/qf0r541rV9Q1+8kvdUvJLqeQ5Ysf0qAg9v170wxk1tToKC0FKpchIwMAYFNNT+XTGj9K2sRcgamdD0qwV9qi24aiwDXXcprP+aGf2NbCpuFU7235zim11EmWLO5ZSBmt2GTKhie3NcrBIsT72bCDvTbq+n1HMEOVh6HH8X1oVSyE43NDWfF8dqTFYoJ5enmH7i/T1qvo+ptdI0t55jSfoPoOgqjFpIU/vM1qLZAQh4RmRRyP7wrNe0lK7K91KyNWw1eSKRWsY50lB+VuBg17h4f1n7bFEJhtdlBOfUivn+0Z3aBldcRngHgkc8V6n4N8SQ3TQW+cXEeFKtwTivEzuhKpTVRLb8j1sqrRjNwb3PVFtgVB9RUclqByTgUkF/GY+WAbuCawvFXi6z0KwaWRwztkRxg8sf8K+Xp0pTkoQV2z3ZVFFOUnohviDXdP0RV86X53+6o6n3+leNeK/Gkt5PIXcttJ2onRRVLxD4ivdbvXuJ5MZ4VR0VfSueniaXjog5PvX2OAy2OGjzPWf5HzeMxsq0uVaRMq71DU7i53RXM0W8/KiSNgfrVwz69ZhSt7JIMZw/zfzq3pmnia481hhV6VpXaAL0r0I0W9WzhlNbGN/wkV28ZjuolyRgMoxUukwHaZGHXmpWtEmKllBxWgsQihwBitIQle8mTKS6FOUYenNHlQRRImWzU8SfJirsTciUZ4pnlYNS4w1SFOQw+lPlC5AE/OmSZGKu7BjionjDGk4jUi3pD/wCiAejH+dFR6YR9nP8AvH+dFax2M3uTIoNOaEkcUmdpVhyDzVmMhqzRZRZMUzGKvzRZ6VUeIg0wuQlM81HIuBwKsAH0oKA0h3K8b7adKvnL0p/lAHmpY0C/SnYlmNLZrIdrDvVy3gSIBUAFWJof4gKjQYNJRVx3HNGGFOgfym6cd6copDH6VVhEk9sozPEMo3LqPX1FEUl1BtuLaXdt5AJwy/Q0+3kMeAeV9KbMnkvviOA3OKTimrME2tjtND+IzMNl+redt2h1H3sDvXMeINXl1S+aeY5ZvuRjoorPtwcvMq/O/QdgPWlEW3JOWc964qGAo0qjqQWrOqrjKtSChJ6ECwliXfv0qGWMykRJ3q+6Hp2qS3twh3Y5rt5Tl5gt7dbeEL3qndZZsVpTMNpqisZd81VtLE3GwQdz2qSU7uKkb5F4qMDPJoHcgKU6Jeam2UsaUWC5XKc5qTZlKcVxx71Ii5GKdhXIlGRTQPnNS7SpIOKZj95SsCG6dxAf94/zNFFrhYzn+8f5mimmBaC4MkR6xnj6HpToWKnmklcefBcDlZf3bfj0/WnMhVzis0WWRhxio5YAecUsZqVulWSUTFg9KCnfvVthxjvUbLRYLlcx+tNZcVYIFMKg9aAuQnkYqGSMqcjpVkx0bQRigCuh7VKADR5PNOCgcU7CuN29ajusiE47Cp8c4ptyB5TAjORSsCJIHWBQpVsBQcj6dOvWo4m8xicEAnOD25ozujQ+qj+VEfBNJR6lud1YnCDNP6UlIauxncbLluKRUC8U/jvSUAMZe1NEdSqhkYAAkngAV3fhv4Q67rOyS9C6ZAw3Zm5kK9SdvYf72KidSMFeTGk3scAI+1IqHdjFdxrMng/w87Wmj2j63dJ8rXl458kN/sIuN348fWuTubmS7lMspTceMIgVQPQAAAUQk5a20CSsU2Xk04LinEcmkrSwhrioW4dfrUkrYqLdkjNJgVLiYQtsB7k/rRUW0STSs4z83FFZXZdjSZSnmWxOM/PGfccir2RLGkg/iANUtQBKiRfvpyKn02QTWgxzgn8KFuHQmXg1L1FMAAp2a0JuIRmkIFOJppoFcjIppFPbr1puM0BcjIweaTHSnt1o60wGgYPNJjvTqQ8igQ33qOX5lalk+VeOtNY4j+lAxU+aCI/7IpR3+lJH/wAe8R7Y/rShQaAZPnpSbs0xSMD3FKTQA4HJp1MU8mnjtTA1/C2uL4c1u31JrWO58nOEfsSCNw9x1H0r1U/Gzw9M7wzaXfmzntzBMuV3AknJBzznPP0FeK4pcCsamHhUd5FRm47G1rieFmZ5NFutVUfww3UCH8N4f+lYJJAp3FNcgdK0jHlViZO41mphagtnIqBpMVVwFkbIqKSTaOKZJNtYenemXjeXEX7BSahvQaRUilVtxJ70VWtQ8cfzxiSQ8sM/d9BRWSehodFO2Y949Oah0WTZLPH2OGFSuAyMgPDDis/S5tt6VPBwVNN6MnodBnmgnikGAKTNakDs0hNNzgZppbntQA4/1pGGOlIGAHNIzd6YhrHByaO1IxzTc880wHGkz700vxTS3NAwk5phOVK9zTieKjz8y470mIkT/j1h+h/nTh0qN0JjjG7AC9qWIBWUF88/3qRVhd2FXFO3U1wFOMYx0pM80xEqmpN2MVEuMdaeMY61QEgbNDN6UwEetNeUDA9aBD93HSombmnb1x1qJmFK4EbthqilODn1pZsgZquX3xnHUc1DZSQyZsgioLq6b7Ptxk9BRPJ8oYdDVRn82eNey/Mayky0jQtokhiAYkseWPvRTVk4+6TRTVibMnLyWLhXJaLsT2qrdN9lv47lD+7fGT6GtEeXcxleo9D2rIut1tmOQZj7H0qZ6FLU6qJw8asOaJH21yel65JDKElc7CSMHtW80/mc5q4VFJXQpRsWGl7CkViearLIAalVq0RDJS59KQODzUZfjimBxnrTuBY688U1yOKjLhR1phmzTuIkLYAqPdyaRnyKYT3pNjJA3qaY7YwQehpm7FLnINTcC2DG6rlRxSvchRgVWDfLTWOfemFyXzPNYmlzUKfL+NPBFAEyninA8VXEnpTvMoAnBFIwzzTA4zRuyetO4AeKaxxSsaibNIBsjZHNUy3lye1WmNVZ14zUSKRTuZ0hBRuh5U1BYASs0vXJwtRazk2hlH3k4/A03Trsx26LEu6U9Pb3rmc/fszVLQ2d8FuAJpArHtRUENsuC037yRuSTRW2vYzNCe3w25Dg1UuSZUKS4PvV+brVK7QGNuKc0CZzN3GYJOOmcg10NhcGa2jYnkqKxXYysUcAitDTSQir2Armp6SNZbG1CQalLVBD92n55rsRgx+7GaYWprk4puTim2Kw6STC9ah83nrTZiahBOahspIurJkUrNxUEZOKexqhMCadE244qCQnBp9mc5zSvqFiz0HWmk80r0hGKpiAE0pam0jd6Q7DfNwacJQaqk/OaetTcqxaWQ04PUK0r/dzTuTYm8zJprNUCsc1JnjNO4WEduKryHIqRyagkPFS2NIo3aKylGGVcYNQaLEiw5YknOKsXHIqDSCTAvNc/wBtGv2TYTyccRs1FNVmx99h+NFb3Isf/9k=)\n",
        "\n",
        "PhD, ARS  \n",
        "Scientist\n",
        "\n",
        "School of Bioinformatics\n",
        "\n",
        "ICAR-Indian Institute of Agricultural Biotechnology, Ranchi, Jharkhand.\n",
        "\n",
        "Spy on me at :)  \n",
        "[Google Scholar](https://scholar.google.com/citations?hl=en&user=0dQ7Sf8AAAAJ&view_op=list_works);    [ResearchGate](https://www.researchgate.net/profile/Kanaka-K-K/research);    [Website](https://iiab.icar.gov.in/staff/dr-kanaka-k-k/)"
      ],
      "metadata": {
        "id": "YqmYkeW7DSqC"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**NGS Variant Calling Tutorial**\n",
        "\n",
        "- **Aim:** Identify genetic variants (SNPs) in E. coli samples\n",
        "- **Tools:** fastp, BWA, SAMtools, BCFtools\n",
        "- **Input:** Two E. coli K-12 paired-end sequencing samples (SRR10527348, SRR10527349)\n",
        "- **Steps:**\n",
        "  1. Quality control of raw reads\n",
        "  2. Alignment to reference genome\n",
        "  3. BAM file processing\n",
        "  4. Variant calling and filtering\n",
        "- **Environment:** Google Colab (cloud-based, no local installation needed)\n",
        "- **Expected output:** VCF file containing high-quality genetic variants\n",
        "- **Time:** ~2 hours (depends on data download speed)\n",
        "\n",
        "# If you find descriptions as boring, skip the text and simply run the codes!"
      ],
      "metadata": {
        "id": "jn5q7u_uDNRJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# i. Setup Environment\n",
        "\"\"\"First, we need to install the required bioinformatics tools\"\"\""
      ],
      "metadata": {
        "id": "-T4AAjzaw7_f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Import required libraries and setup directories\n",
        "import os\n",
        "from google.colab import drive\n",
        "import datetime\n",
        "\n",
        "# Mount Google Drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "# Create directory structure\n",
        "project_dir = \"/content/variant_calling\"\n",
        "os.makedirs(project_dir, exist_ok=True)\n",
        "\n",
        "# Create subdirectories\n",
        "dirs = [\n",
        "    'raw_data',\n",
        "    'cleaned_data',\n",
        "    'reference',\n",
        "    'alignments',\n",
        "    'variants',\n",
        "    'reports'\n",
        "]\n",
        "\n",
        "for d in dirs:\n",
        "    os.makedirs(f\"{project_dir}/{d}\", exist_ok=True)\n",
        "\n",
        "print(\"Directory structure created successfully!\")"
      ],
      "metadata": {
        "id": "R5DD4TH07Cze",
        "outputId": "11ea075c-0b88-415c-a84b-b9c8eeb2bd2f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n",
            "Directory structure created successfully!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# ii. Install tools"
      ],
      "metadata": {
        "id": "uGICnduL7SHr"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**fastp**\n",
        "- Quality control tool for FASTQ files\n",
        "- Removes low quality reads/bases\n",
        "- Trims sequencing adapters\n",
        "- Provides QC reports with statistics and graphs\n",
        "- Filters out contaminated reads\n",
        "\n",
        "**BWA (Burrows-Wheeler Aligner)**\n",
        "- Aligns sequencing reads to reference genome\n",
        "- Specially designed for short read alignment\n",
        "- BWA-MEM algorithm handles paired-end reads\n",
        "- Outputs alignments in SAM format\n",
        "- Handles mismatches and gaps in sequences\n",
        "\n",
        "**SAMtools**\n",
        "- Manipulates SAM/BAM format files\n",
        "- Converts SAM to BAM (binary format)\n",
        "- Sorts and indexes BAM files\n",
        "- Generates alignment statistics\n",
        "- Extracts specific regions from alignments\n",
        "- Views and filters aligned reads\n",
        "\n",
        "**BCFtools**\n",
        "- Variant calling from BAM files\n",
        "- Creates and manipulates VCF files\n",
        "- Filters variants based on quality scores\n",
        "- Performs variant genotyping\n",
        "- Handles multi-sample variant calling\n",
        "- Allows annotation of variants"
      ],
      "metadata": {
        "id": "AmttWkFoE53F"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install required tools\n",
        "!apt-get update && apt-get install -y fastp bwa samtools bcftools"
      ],
      "metadata": {
        "id": "DW-cM6c47Ucn",
        "collapsed": true,
        "outputId": "39264376-6a19-4633-b980-47f30365697b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Get:1 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Hit:2 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Get:4 https://cli.github.com/packages stable InRelease [3,917 B]\n",
            "Get:5 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Get:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease [18.1 kB]\n",
            "Get:9 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [3,677 kB]\n",
            "Get:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease [24.6 kB]\n",
            "Get:11 https://cli.github.com/packages stable/main amd64 Packages [356 B]\n",
            "Get:12 http://security.ubuntu.com/ubuntu jammy-security/multiverse amd64 Packages [62.6 kB]\n",
            "Get:13 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [6,388 kB]\n",
            "Get:14 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,297 kB]\n",
            "Get:15 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages [83.8 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [70.9 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [4,014 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,608 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [6,607 kB]\n",
            "Get:20 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,712 kB]\n",
            "Get:21 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,892 kB]\n",
            "Get:22 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy/main amd64 Packages [38.8 kB]\n",
            "Get:23 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy/main amd64 Packages [75.3 kB]\n",
            "Fetched 37.0 MB in 4s (8,235 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libhts3 libhtscodecs2\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bcftools bwa fastp libhts3 libhtscodecs2 samtools\n",
            "0 upgraded, 6 newly installed, 0 to remove and 53 not upgraded.\n",
            "Need to get 2,048 kB of archives.\n",
            "After this operation, 5,736 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bwa amd64 0.7.17-6 [195 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fastp amd64 0.20.1+dfsg-1 [193 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 2,048 kB in 0s (7,921 kB/s)\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 117540 files and directories currently installed.)\n",
            "Preparing to unpack .../0-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../1-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package bcftools.\n",
            "Preparing to unpack .../2-bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Selecting previously unselected package bwa.\n",
            "Preparing to unpack .../3-bwa_0.7.17-6_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-6) ...\n",
            "Selecting previously unselected package fastp.\n",
            "Preparing to unpack .../4-fastp_0.20.1+dfsg-1_amd64.deb ...\n",
            "Unpacking fastp (0.20.1+dfsg-1) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../5-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up bwa (0.7.17-6) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up fastp (0.20.1+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.11) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install SRA toolkit\n",
        "!wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz\n",
        "!tar -xzf sratoolkit.current-ubuntu64.tar.gz\n",
        "!mv sratoolkit.*/bin/* /usr/local/bin/\n",
        "!rm sratoolkit.current-ubuntu64.tar.gz"
      ],
      "metadata": {
        "id": "hP5U31QPA-DR",
        "outputId": "082aed42-c350-423d-ecf7-bc83cf656cde",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2026-02-08 04:58:44--  https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz\n",
            "Resolving ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.13, 130.14.250.31, ...\n",
            "Connecting to ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 92046842 (88M) [application/x-gzip]\n",
            "Saving to: ‘sratoolkit.current-ubuntu64.tar.gz’\n",
            "\n",
            "sratoolkit.current- 100%[===================>]  87.78M   130MB/s    in 0.7s    \n",
            "\n",
            "2026-02-08 04:58:45 (130 MB/s) - ‘sratoolkit.current-ubuntu64.tar.gz’ saved [92046842/92046842]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# iii. Understanding the Input Data: FASTQ Format\n",
        "\n",
        "FASTQ files contain sequencing reads with quality scores. Each read has 4 lines:\n",
        "1. Read identifier (starts with @)\n",
        "2. Sequence\n",
        "3. '+' separator\n",
        "4. Quality scores (ASCII-encoded)\n",
        "\n",
        "Example of FASTQ format:\n",
        "\n",
        "@SRR001666.1\n",
        "\n",
        "GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGT\n",
        "\n",
        "'+'\n",
        "\n",
        "!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF>>>>>>CCCCCCC65\n",
        "\n",
        "Let's download a small test dataset (E. coli K-12): This tutorial demonstrates variant calling with two samples, each with paired-end reads.\n",
        "Sample1 and Sample2\n",
        "Each sample has forward (R1) and reverse (R2) reads."
      ],
      "metadata": {
        "id": "ZNK8WQh5x80n"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Download samples"
      ],
      "metadata": {
        "id": "KuZn053PFEXJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Download sample 1\n",
        "%cd /content/variant_calling/raw_data\n",
        "print(\"Downloading Sample 1...\")\n",
        "!fastq-dump --split-files SRR10527348\n",
        "print(\"\\nSample 1 files:\")\n",
        "!ls -lh SRR10527348*"
      ],
      "metadata": {
        "id": "c3Dn6Al9yMlK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f501afed-b376-4f67-bb38-969e5a352bd8"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/variant_calling/raw_data\n",
            "Downloading Sample 1...\n",
            "Read 2281618 spots for SRR10527348\n",
            "Written 2281618 spots for SRR10527348\n",
            "\n",
            "Sample 1 files:\n",
            "-rw-r--r-- 1 root root 609M Feb  8 05:04 SRR10527348_1.fastq\n",
            "-rw-r--r-- 1 root root 609M Feb  8 05:04 SRR10527348_2.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download sample 2\n",
        "%cd /content/variant_calling/raw_data\n",
        "print(\"\\nDownloading Sample 2...\")\n",
        "!fastq-dump --split-files SRR10527349\n",
        "print(\"\\nSample 2 files:\")\n",
        "!ls -lh SRR10527349*"
      ],
      "metadata": {
        "id": "wJs72Tw7BG2N",
        "outputId": "867a05e1-7c84-4a43-a507-1e61caef2616",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/variant_calling/raw_data\n",
            "\n",
            "Downloading Sample 2...\n",
            "Read 1649218 spots for SRR10527349\n",
            "Written 1649218 spots for SRR10527349\n",
            "\n",
            "Sample 2 files:\n",
            "-rw-r--r-- 1 root root 440M Feb  8 05:06 SRR10527349_1.fastq\n",
            "-rw-r--r-- 1 root root 440M Feb  8 05:06 SRR10527349_2.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Reference Genome\n",
        "**Reference Genome**\n",
        "\n",
        "**Definition**: A digital nucleotide sequence database assembled to represent a species' genetic makeup, used as a standard for comparison in sequencing studies.\n",
        "\n",
        "**Common Formats**:\n",
        "- FASTA (.fa/.fasta)\n",
        "  - Simple text format\n",
        "  - Header line starts with '>' symbol\n",
        "  - Contains sequence data in subsequent lines\n",
        "  - Example:\n",
        "    ```\n",
        "    >Chr1\n",
        "    ATGCTAGCTAGCTGATCG\n",
        "    GCTAGCTAGCTGATCGAT\n",
        "    ```\n",
        "\n",
        "**Required Files for Alignment**:\n",
        "- Main reference file (.fasta)\n",
        "- Index files (created by tools)\n",
        "  - BWA index (.amb, .ann, .bwt, .pac, .sa)\n",
        "  - FASTA index (.fai)\n",
        "  - Dictionary (.dict)\n",
        "\n",
        "**Common Sources**:\n",
        "- NCBI RefSeq\n",
        "- Ensembl\n",
        "- UCSC Genome Browser\n",
        "\n",
        "**E. coli Reference Example**:\n",
        "- Strain: K-12 MG1655\n",
        "- Size: ~4.6 Mb\n",
        "- Single circular chromosome\n",
        "- Well-annotated for teaching purposes"
      ],
      "metadata": {
        "id": "9LPK8yX1zXTH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Download the reference genome\n",
        "%cd /content/variant_calling/reference\n",
        "\n",
        "print(\"Downloading reference genome...\")\n",
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "!gunzip GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "!mv GCF_000005845.2_ASM584v2_genomic.fna reference.fasta"
      ],
      "metadata": {
        "id": "xTnUAF4aze8A",
        "outputId": "2d010dd9-b37d-4ae6-cd4c-df42303132d5",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/variant_calling/reference\n",
            "Downloading reference genome...\n",
            "--2026-02-08 05:07:30--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.31, 130.14.250.7, 130.14.250.10, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.31|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1379902 (1.3M) [application/x-gzip]\n",
            "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’\n",
            "\n",
            "GCF_000005845.2_ASM 100%[===================>]   1.32M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2026-02-08 05:07:30 (14.6 MB/s) - ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’ saved [1379902/1379902]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Why Index Reference Files**\n",
        "\n",
        "1. **Speed & Efficiency**\n",
        "   - Enables quick lookup of sequences\n",
        "   - Avoids scanning entire file for each query\n",
        "   - Reduces memory usage\n",
        "\n",
        "2. **Random Access**\n",
        "   - Allows direct access to specific regions\n",
        "   - Skips reading entire genome\n",
        "   - Essential for targeted analysis\n",
        "\n",
        "3. **Tool Requirements**\n",
        "   - BWA requires .amb, .ann, .bwt, .pac, .sa files\n",
        "   - SAMtools needs .fai file\n",
        "   - Most alignment tools won't work without indexes\n",
        "\n",
        "4. **Common Analogy**\n",
        "   - Like a book's table of contents\n",
        "   - Helps find content quickly\n",
        "   - Instead of reading every page\n",
        "\n",
        "5. **Practical Impact**\n",
        "   - Reduces runtime from hours to minutes\n",
        "   - Enables processing of large datasets\n",
        "   - Critical for high-throughput analysis"
      ],
      "metadata": {
        "id": "DzIiCyKFFhVU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Index reference genome\n",
        "print(\"\\nIndexing reference genome...\")\n",
        "!bwa index reference.fasta\n",
        "!samtools faidx reference.fasta"
      ],
      "metadata": {
        "id": "NqdL11q_BgYH",
        "outputId": "ad5c8dd5-17c7-4882-8d00-ece79e19e648",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Indexing reference genome...\n",
            "[bwa_index] Pack FASTA... 0.04 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 2.28 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.04 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.04 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 1.39 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index reference.fasta\n",
            "[main] Real time: 3.981 sec; CPU: 3.797 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Quality Control with fastp\n",
        "\n",
        "fastp performs quality control and filtering on FASTQ files.\n",
        "- Removes low-quality bases\n",
        "- Trims adapters\n",
        "- Generates QC reports\n",
        "\n",
        "Input:\n",
        "- Raw FASTQ files\n",
        "\n",
        "\n",
        "Output:\n",
        "- Cleaned FASTQ files\n",
        "- HTML report with quality metrics\n",
        "\n",
        "**FASTQ Format** (Cleaned FASTQ files)\n",
        "```\n",
        "@SRR10527348.1\n",
        "ATGCTAGCTAGCTGATCGATGC\n",
        "+\n",
        "!''*((((***+))%%%++)(%%\n",
        "```\n",
        "- Line 1: Header (starts with @)\n",
        "- Line 2: DNA sequence\n",
        "- Line 3: Separator (+)\n",
        "- Line 4: Quality scores (ASCII-encoded)\n",
        "\n",
        "**HTML Report Format**\n",
        "```html\n",
        "Summary Section:\n",
        "- Total reads: 1,000,000\n",
        "- Filtered reads: 50,000\n",
        "- Q30 bases: 95%\n",
        "\n",
        "Graphs/Charts:\n",
        "- Quality score distribution\n",
        "- GC content plot\n",
        "- Length distribution\n",
        "- Adapter content\n",
        "\n",
        "Detailed Statistics:\n",
        "- Before/After filtering metrics\n",
        "- Duplication rates\n",
        "- Base composition\n",
        "```\n",
        "\n",
        "Key Features of HTML Report:\n",
        "- Interactive visualizations\n",
        "- Color-coded metrics\n",
        "- Downloadable charts\n",
        "- Easy to share/view in browser\n",
        "- No special software needed\n",
        "\n",
        "*Note: fastp generates comprehensive reports that help assess sequencing quality and effectiveness of cleaning steps*"
      ],
      "metadata": {
        "id": "bE9qFAvczr0n"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Check number of lines in each sample\n",
        "!wc -l \\\n",
        "/content/variant_calling/raw_data/SRR10527348_1.fastq \\\n",
        "/content/variant_calling/raw_data/SRR10527348_2.fastq \\\n",
        "/content/variant_calling/raw_data/SRR10527349_1.fastq \\\n",
        "/content/variant_calling/raw_data/SRR10527349_2.fastq"
      ],
      "metadata": {
        "id": "XYEKSqqkp6HT",
        "outputId": "fd751c02-e515-41e5-9da6-325d87acde1c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   9126472 /content/variant_calling/raw_data/SRR10527348_1.fastq\n",
            "   9126472 /content/variant_calling/raw_data/SRR10527348_2.fastq\n",
            "   6596872 /content/variant_calling/raw_data/SRR10527349_1.fastq\n",
            "   6596872 /content/variant_calling/raw_data/SRR10527349_2.fastq\n",
            "  31446688 total\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check number of reads. Reads = (Number of lines/4)\n",
        "!awk '{printf \"%s\\t%d reads\\n\", $2, $1/4}' \\\n",
        "<(wc -l /content/variant_calling/raw_data/*.fastq)"
      ],
      "metadata": {
        "id": "nFIm9aB-qfhj",
        "outputId": "b7cc0de8-e84e-4523-a282-c51db6faabe7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/variant_calling/raw_data/SRR10527348_1.fastq\t2281618 reads\n",
            "/content/variant_calling/raw_data/SRR10527348_2.fastq\t2281618 reads\n",
            "/content/variant_calling/raw_data/SRR10527349_1.fastq\t1649218 reads\n",
            "/content/variant_calling/raw_data/SRR10527349_2.fastq\t1649218 reads\n",
            "total\t7861672 reads\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Process Sample 1\n",
        "print(\"Processing Sample 1...\")\n",
        "!fastp \\\n",
        "    -i /content/variant_calling/raw_data/SRR10527348_1.fastq \\\n",
        "    -I /content/variant_calling/raw_data/SRR10527348_2.fastq \\\n",
        "    -o /content/variant_calling/cleaned_data/SRR10527348_1.clean.fastq.gz \\\n",
        "    -O /content/variant_calling/cleaned_data/SRR10527348_2.clean.fastq.gz \\\n",
        "    --html /content/variant_calling/reports/SRR10527348_fastp.html \\\n",
        "    --json /content/variant_calling/reports/SRR10527348_fastp.json \\\n",
        "    --qualified_quality_phred 20 \\\n",
        "    --length_required 50"
      ],
      "metadata": {
        "collapsed": true,
        "id": "9I5srfCVzxsB",
        "outputId": "5c549547-6ce1-4c7d-f384-965acfb07180",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing Sample 1...\n",
            "Read1 before filtering:\n",
            "total reads: 2281618\n",
            "total bases: 228128864\n",
            "Q20 bases: 223900507(98.1465%)\n",
            "Q30 bases: 220104460(96.4825%)\n",
            "\n",
            "Read2 before filtering:\n",
            "total reads: 2281618\n",
            "total bases: 228137436\n",
            "Q20 bases: 215029047(94.2542%)\n",
            "Q30 bases: 206512369(90.521%)\n",
            "\n",
            "Read1 after filtering:\n",
            "total reads: 2167596\n",
            "total bases: 216989871\n",
            "Q20 bases: 214258426(98.7412%)\n",
            "Q30 bases: 211636486(97.5329%)\n",
            "\n",
            "Read2 aftering filtering:\n",
            "total reads: 2167596\n",
            "total bases: 216987041\n",
            "Q20 bases: 209884678(96.7268%)\n",
            "Q30 bases: 204157255(94.0873%)\n",
            "\n",
            "Filtering result:\n",
            "reads passed filter: 4335192\n",
            "reads failed due to low quality: 216638\n",
            "reads failed due to too many N: 158\n",
            "reads failed due to too short: 11248\n",
            "reads with adapter trimmed: 2012\n",
            "bases trimmed due to adapters: 33402\n",
            "\n",
            "Duplication rate: 1.21472%\n",
            "\n",
            "Insert size peak (evaluated by paired-end reads): 166\n",
            "\n",
            "JSON report: /content/variant_calling/reports/SRR10527348_fastp.json\n",
            "HTML report: /content/variant_calling/reports/SRR10527348_fastp.html\n",
            "\n",
            "fastp -i /content/variant_calling/raw_data/SRR10527348_1.fastq -I /content/variant_calling/raw_data/SRR10527348_2.fastq -o /content/variant_calling/cleaned_data/SRR10527348_1.clean.fastq.gz -O /content/variant_calling/cleaned_data/SRR10527348_2.clean.fastq.gz --html /content/variant_calling/reports/SRR10527348_fastp.html --json /content/variant_calling/reports/SRR10527348_fastp.json --qualified_quality_phred 20 --length_required 50 \n",
            "fastp v0.20.1, time used: 70 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Process Sample 2\n",
        "print(\"\\nProcessing Sample 2...\")\n",
        "!fastp \\\n",
        "    -i /content/variant_calling/raw_data/SRR10527349_1.fastq \\\n",
        "    -I /content/variant_calling/raw_data/SRR10527349_2.fastq \\\n",
        "    -o /content/variant_calling/cleaned_data/SRR10527349_1.clean.fastq.gz \\\n",
        "    -O /content/variant_calling/cleaned_data/SRR10527349_2.clean.fastq.gz \\\n",
        "    --html /content/variant_calling/reports/SRR10527349_fastp.html \\\n",
        "    --json /content/variant_calling/reports/SRR10527349_fastp.json \\\n",
        "    --qualified_quality_phred 20 \\\n",
        "    --length_required 50"
      ],
      "metadata": {
        "id": "XlL9WPKSBoJh",
        "collapsed": true,
        "outputId": "7d22649b-4b28-4e15-9bba-9eba2ae237d0",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Processing Sample 2...\n",
            "Read1 before filtering:\n",
            "total reads: 1649218\n",
            "total bases: 165347518\n",
            "Q20 bases: 162232943(98.1163%)\n",
            "Q30 bases: 159457249(96.4376%)\n",
            "\n",
            "Read2 before filtering:\n",
            "total reads: 1649218\n",
            "total bases: 165351647\n",
            "Q20 bases: 153954738(93.1075%)\n",
            "Q30 bases: 146686707(88.712%)\n",
            "\n",
            "Read1 after filtering:\n",
            "total reads: 1547306\n",
            "total bases: 155228895\n",
            "Q20 bases: 153434087(98.8438%)\n",
            "Q30 bases: 151699724(97.7265%)\n",
            "\n",
            "Read2 aftering filtering:\n",
            "total reads: 1547306\n",
            "total bases: 155226992\n",
            "Q20 bases: 149349122(96.2134%)\n",
            "Q30 bases: 144670098(93.1991%)\n",
            "\n",
            "Filtering result:\n",
            "reads passed filter: 3094612\n",
            "reads failed due to low quality: 199102\n",
            "reads failed due to too many N: 100\n",
            "reads failed due to too short: 4622\n",
            "reads with adapter trimmed: 1036\n",
            "bases trimmed due to adapters: 22308\n",
            "\n",
            "Duplication rate: 0.989916%\n",
            "\n",
            "Insert size peak (evaluated by paired-end reads): 167\n",
            "\n",
            "JSON report: /content/variant_calling/reports/SRR10527349_fastp.json\n",
            "HTML report: /content/variant_calling/reports/SRR10527349_fastp.html\n",
            "\n",
            "fastp -i /content/variant_calling/raw_data/SRR10527349_1.fastq -I /content/variant_calling/raw_data/SRR10527349_2.fastq -o /content/variant_calling/cleaned_data/SRR10527349_1.clean.fastq.gz -O /content/variant_calling/cleaned_data/SRR10527349_2.clean.fastq.gz --html /content/variant_calling/reports/SRR10527349_fastp.html --json /content/variant_calling/reports/SRR10527349_fastp.json --qualified_quality_phred 20 --length_required 50 \n",
            "fastp v0.20.1, time used: 56 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 5. Alignment with BWA-MEM\n",
        "\n",
        "BWA-MEM aligns cleaned reads to the reference genome.\n",
        "\n",
        "Input: Cleaned FASTQ files + indexed reference\n",
        "\n",
        "Output: SAM file (Sequence Alignment/Map format)\n",
        "\n",
        "Example SAM format:\n",
        "@SQ     SN:ref  LN:45\n",
        "r001    163     ref     7       30      8M2I4M1D3M      =       37      39      TTAGATAAAGGATACTG     *\n"
      ],
      "metadata": {
        "id": "le21w-7Y2zCG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Align Sample 1\n",
        "print(\"Aligning Sample 1...\")\n",
        "!bwa mem -t 2 \\\n",
        "    /content/variant_calling/reference/reference.fasta \\\n",
        "    /content/variant_calling/cleaned_data/SRR10527348_1.clean.fastq.gz \\\n",
        "    /content/variant_calling/cleaned_data/SRR10527348_2.clean.fastq.gz \\\n",
        "    > /content/variant_calling/alignments/SRR10527348.sam"
      ],
      "metadata": {
        "id": "9adeJf7n24RT",
        "outputId": "13ca4cd4-794c-451d-8eb9-4ec86046571e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Aligning Sample 1...\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 199792 sequences (20000167 bp)...\n",
            "[M::process] read 199778 sequences (20000094 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 81836, 40, 13)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (162, 325, 506)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1194)\n",
            "[M::mem_pestat] mean and std.dev: (358.36, 312.59)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1609)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (211, 294, 424)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 850)\n",
            "[M::mem_pestat] mean and std.dev: (326.70, 158.83)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1063)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (120, 244, 786)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2118)\n",
            "[M::mem_pestat] mean and std.dev: (453.81, 532.19)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2784)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (100, 234, 322)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 766)\n",
            "[M::mem_pestat] mean and std.dev: (205.25, 102.82)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 988)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199792 reads in 24.715 CPU sec, 15.743 real sec\n",
            "[M::process] read 199780 sequences (20000001 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 81793, 37, 17)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (296, 356, 419)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (50, 665)\n",
            "[M::mem_pestat] mean and std.dev: (327.10, 132.20)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 856)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (211, 294, 425)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 853)\n",
            "[M::mem_pestat] mean and std.dev: (327.29, 160.12)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1067)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (103, 266, 1006)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2812)\n",
            "[M::mem_pestat] mean and std.dev: (397.42, 397.07)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3715)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 236, 368)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 806)\n",
            "[M::mem_pestat] mean and std.dev: (287.18, 194.63)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1066)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199778 reads in 26.003 CPU sec, 16.480 real sec\n",
            "[M::process] read 199826 sequences (20000037 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (17, 81829, 38, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (121, 330, 668)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1762)\n",
            "[M::mem_pestat] mean and std.dev: (439.29, 424.26)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2309)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (212, 296, 427)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 857)\n",
            "[M::mem_pestat] mean and std.dev: (329.42, 160.81)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1072)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (66, 255, 809)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2295)\n",
            "[M::mem_pestat] mean and std.dev: (465.53, 541.47)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3038)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (183, 308, 572)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1350)\n",
            "[M::mem_pestat] mean and std.dev: (310.50, 223.63)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1739)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199780 reads in 25.337 CPU sec, 15.927 real sec\n",
            "[M::process] read 199868 sequences (20000163 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 81926, 50, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (147, 217, 493)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1185)\n",
            "[M::mem_pestat] mean and std.dev: (288.92, 199.88)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1531)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (211, 292, 423)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 847)\n",
            "[M::mem_pestat] mean and std.dev: (325.95, 158.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1059)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (106, 206, 761)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2071)\n",
            "[M::mem_pestat] mean and std.dev: (394.65, 475.74)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2726)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (228, 318, 1004)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2556)\n",
            "[M::mem_pestat] mean and std.dev: (365.10, 321.67)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3332)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199826 reads in 25.493 CPU sec, 17.158 real sec\n",
            "[M::process] read 199864 sequences (20000039 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (21, 81778, 26, 15)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (123, 213, 490)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1224)\n",
            "[M::mem_pestat] mean and std.dev: (251.05, 177.04)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1591)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 286, 409)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 813)\n",
            "[M::mem_pestat] mean and std.dev: (316.36, 150.73)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1015)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (78, 204, 580)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1584)\n",
            "[M::mem_pestat] mean and std.dev: (297.68, 269.97)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2086)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (213, 335, 438)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 888)\n",
            "[M::mem_pestat] mean and std.dev: (298.79, 133.25)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1113)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199868 reads in 23.437 CPU sec, 14.513 real sec\n",
            "[M::process] read 199680 sequences (20000121 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (14, 81800, 45, 6)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (163, 258, 638)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1588)\n",
            "[M::mem_pestat] mean and std.dev: (396.85, 370.68)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2063)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (206, 284, 405)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 803)\n",
            "[M::mem_pestat] mean and std.dev: (313.27, 148.70)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1002)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (142, 216, 745)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1951)\n",
            "[M::mem_pestat] mean and std.dev: (380.95, 457.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2554)\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_process_seqs] Processed 199864 reads in 23.988 CPU sec, 14.892 real sec\n",
            "[M::process] read 199784 sequences (20000069 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 81833, 33, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (185, 425, 496)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1118)\n",
            "[M::mem_pestat] mean and std.dev: (332.73, 187.38)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1429)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (210, 290, 415)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 825)\n",
            "[M::mem_pestat] mean and std.dev: (321.42, 153.61)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1030)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (79, 187, 855)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2407)\n",
            "[M::mem_pestat] mean and std.dev: (429.19, 489.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3183)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (195, 310, 588)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1374)\n",
            "[M::mem_pestat] mean and std.dev: (318.55, 170.33)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1767)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199680 reads in 24.705 CPU sec, 15.566 real sec\n",
            "[M::process] read 199824 sequences (20000176 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 81743, 54, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (160, 368, 5403)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15889)\n",
            "[M::mem_pestat] mean and std.dev: (1946.31, 2460.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 21132)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (211, 292, 421)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 841)\n",
            "[M::mem_pestat] mean and std.dev: (324.35, 156.50)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1051)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (130, 220, 754)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2002)\n",
            "[M::mem_pestat] mean and std.dev: (434.31, 485.59)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2626)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (100, 590, 1257)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3571)\n",
            "[M::mem_pestat] mean and std.dev: (588.50, 532.55)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4728)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199784 reads in 25.908 CPU sec, 17.212 real sec\n",
            "[M::process] read 199876 sequences (20000035 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (18, 81814, 30, 15)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (95, 241, 416)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1058)\n",
            "[M::mem_pestat] mean and std.dev: (212.40, 159.25)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1379)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (208, 287, 411)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 817)\n",
            "[M::mem_pestat] mean and std.dev: (317.97, 152.25)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1020)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (83, 205, 711)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1967)\n",
            "[M::mem_pestat] mean and std.dev: (396.29, 468.16)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2595)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (153, 163, 542)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1320)\n",
            "[M::mem_pestat] mean and std.dev: (213.38, 166.35)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1709)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199824 reads in 30.102 CPU sec, 19.381 real sec\n",
            "[M::process] read 199838 sequences (20000031 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (23, 81876, 34, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (118, 169, 357)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 835)\n",
            "[M::mem_pestat] mean and std.dev: (208.19, 165.72)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1074)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (206, 283, 403)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 797)\n",
            "[M::mem_pestat] mean and std.dev: (312.58, 147.99)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 994)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (97, 221, 679)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1843)\n",
            "[M::mem_pestat] mean and std.dev: (387.81, 412.14)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2425)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (194, 279, 1239)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3329)\n",
            "[M::mem_pestat] mean and std.dev: (469.83, 434.08)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4374)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199876 reads in 27.990 CPU sec, 18.405 real sec\n",
            "[M::process] read 199682 sequences (20000135 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (16, 81949, 38, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (171, 310, 464)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1050)\n",
            "[M::mem_pestat] mean and std.dev: (308.73, 224.79)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1343)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (205, 281, 398)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 784)\n",
            "[M::mem_pestat] mean and std.dev: (309.24, 144.83)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 977)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (74, 170, 629)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1739)\n",
            "[M::mem_pestat] mean and std.dev: (283.38, 345.20)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2294)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (99, 181, 305)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 717)\n",
            "[M::mem_pestat] mean and std.dev: (181.77, 100.87)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 923)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199838 reads in 24.628 CPU sec, 15.305 real sec\n",
            "[M::process] read 199768 sequences (20000121 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20, 81843, 41, 16)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (152, 262, 624)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1568)\n",
            "[M::mem_pestat] mean and std.dev: (315.94, 325.58)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2040)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (212, 293, 423)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 845)\n",
            "[M::mem_pestat] mean and std.dev: (326.66, 158.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1056)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (95, 513, 865)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2405)\n",
            "[M::mem_pestat] mean and std.dev: (464.92, 468.31)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3175)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (189, 348, 1215)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3267)\n",
            "[M::mem_pestat] mean and std.dev: (484.00, 440.97)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4293)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199682 reads in 27.330 CPU sec, 17.855 real sec\n",
            "[M::process] read 199836 sequences (20000101 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (17, 81730, 29, 16)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (124, 396, 771)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2065)\n",
            "[M::mem_pestat] mean and std.dev: (357.27, 268.49)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2712)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (213, 297, 432)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 870)\n",
            "[M::mem_pestat] mean and std.dev: (331.84, 163.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1089)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (124, 440, 2435)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7057)\n",
            "[M::mem_pestat] mean and std.dev: (853.92, 963.30)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9368)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (104, 160, 324)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 764)\n",
            "[M::mem_pestat] mean and std.dev: (201.06, 149.27)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 984)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199768 reads in 24.138 CPU sec, 14.940 real sec\n",
            "[M::process] read 199812 sequences (20000186 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 81992, 38, 6)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (209, 292, 421)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 845)\n",
            "[M::mem_pestat] mean and std.dev: (324.45, 158.37)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1057)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (168, 333, 678)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1698)\n",
            "[M::mem_pestat] mean and std.dev: (397.52, 373.70)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2208)\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_process_seqs] Processed 199836 reads in 24.766 CPU sec, 15.533 real sec\n",
            "[M::process] read 199832 sequences (20000074 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 81869, 46, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (121, 287, 962)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2644)\n",
            "[M::mem_pestat] mean and std.dev: (399.83, 446.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3485)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (209, 288, 413)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 821)\n",
            "[M::mem_pestat] mean and std.dev: (319.48, 153.08)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1025)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (101, 176, 845)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2333)\n",
            "[M::mem_pestat] mean and std.dev: (410.61, 533.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3077)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (119, 213, 367)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 863)\n",
            "[M::mem_pestat] mean and std.dev: (193.30, 97.58)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1111)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199812 reads in 23.061 CPU sec, 14.242 real sec\n",
            "[M::process] read 199694 sequences (20000020 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (19, 81746, 36, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (268, 380, 5491)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15937)\n",
            "[M::mem_pestat] mean and std.dev: (1888.74, 2350.40)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 21160)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (208, 286, 411)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 817)\n",
            "[M::mem_pestat] mean and std.dev: (317.78, 152.09)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1020)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (106, 208, 695)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1873)\n",
            "[M::mem_pestat] mean and std.dev: (426.31, 442.84)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2462)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (104, 288, 449)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1139)\n",
            "[M::mem_pestat] mean and std.dev: (225.90, 140.92)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1484)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199832 reads in 24.049 CPU sec, 14.890 real sec\n",
            "[M::process] read 199736 sequences (20000060 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (18, 81454, 34, 18)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (200, 265, 394)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 782)\n",
            "[M::mem_pestat] mean and std.dev: (262.94, 140.25)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 976)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (213, 297, 431)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 867)\n",
            "[M::mem_pestat] mean and std.dev: (331.51, 162.71)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1085)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (120, 557, 1566)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4458)\n",
            "[M::mem_pestat] mean and std.dev: (1011.94, 1245.54)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5994)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (154, 365, 457)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1063)\n",
            "[M::mem_pestat] mean and std.dev: (308.06, 232.81)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1366)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199694 reads in 23.821 CPU sec, 14.737 real sec\n",
            "[M::process] read 199780 sequences (20000070 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (14, 81835, 43, 21)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (182, 230, 285)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 491)\n",
            "[M::mem_pestat] mean and std.dev: (190.08, 78.72)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 594)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (213, 297, 430)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 864)\n",
            "[M::mem_pestat] mean and std.dev: (331.05, 161.55)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1081)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (127, 170, 763)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2035)\n",
            "[M::mem_pestat] mean and std.dev: (364.80, 432.50)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2671)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (105, 162, 298)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 684)\n",
            "[M::mem_pestat] mean and std.dev: (196.90, 132.93)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 877)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199736 reads in 23.670 CPU sec, 14.739 real sec\n",
            "[M::process] read 199822 sequences (20000141 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 81931, 42, 7)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (100, 280, 465)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1195)\n",
            "[M::mem_pestat] mean and std.dev: (282.31, 178.57)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1560)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (210, 291, 420)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 840)\n",
            "[M::mem_pestat] mean and std.dev: (324.24, 157.15)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1050)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (118, 206, 891)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2437)\n",
            "[M::mem_pestat] mean and std.dev: (428.38, 525.36)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3210)\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_process_seqs] Processed 199780 reads in 22.511 CPU sec, 13.765 real sec\n",
            "[M::process] read 199842 sequences (20000059 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (14, 81876, 31, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (202, 284, 438)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 910)\n",
            "[M::mem_pestat] mean and std.dev: (282.23, 146.23)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1146)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (209, 288, 414)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 824)\n",
            "[M::mem_pestat] mean and std.dev: (319.68, 153.20)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1029)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (93, 262, 753)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2073)\n",
            "[M::mem_pestat] mean and std.dev: (330.18, 293.58)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2733)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (141, 217, 434)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1020)\n",
            "[M::mem_pestat] mean and std.dev: (284.08, 226.72)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1313)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199822 reads in 22.648 CPU sec, 13.841 real sec\n",
            "[M::process] read 139478 sequences (13975012 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (16, 81857, 38, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (157, 195, 340)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 706)\n",
            "[M::mem_pestat] mean and std.dev: (213.00, 111.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 889)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (207, 283, 405)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 801)\n",
            "[M::mem_pestat] mean and std.dev: (314.08, 149.20)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 999)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (88, 197, 811)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2257)\n",
            "[M::mem_pestat] mean and std.dev: (422.54, 485.43)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2980)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (158, 230, 393)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 863)\n",
            "[M::mem_pestat] mean and std.dev: (246.23, 141.63)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1098)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199842 reads in 23.933 CPU sec, 15.760 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 57185, 27, 11)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (148, 232, 885)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2359)\n",
            "[M::mem_pestat] mean and std.dev: (253.45, 215.86)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3096)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (209, 286, 408)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 806)\n",
            "[M::mem_pestat] mean and std.dev: (316.72, 149.03)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1005)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (151, 207, 848)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2242)\n",
            "[M::mem_pestat] mean and std.dev: (385.50, 392.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2939)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (269, 319, 1271)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3275)\n",
            "[M::mem_pestat] mean and std.dev: (393.67, 364.62)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4277)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 139478 reads in 16.668 CPU sec, 10.700 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem -t 2 /content/variant_calling/reference/reference.fasta /content/variant_calling/cleaned_data/SRR10527348_1.clean.fastq.gz /content/variant_calling/cleaned_data/SRR10527348_2.clean.fastq.gz\n",
            "[main] Real time: 342.482 sec; CPU: 539.633 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Align Sample 2\n",
        "print(\"\\nAligning Sample 2...\")\n",
        "!bwa mem -t 2 \\\n",
        "    /content/variant_calling/reference/reference.fasta \\\n",
        "    /content/variant_calling/cleaned_data/SRR10527349_1.clean.fastq.gz \\\n",
        "    /content/variant_calling/cleaned_data/SRR10527349_2.clean.fastq.gz \\\n",
        "    > /content/variant_calling/alignments/SRR10527349.sam"
      ],
      "metadata": {
        "id": "-Ix0mKGeBvaC",
        "outputId": "08df27c6-7b82-47c3-e36a-14d4394450bd",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Aligning Sample 2...\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 199314 sequences (20000048 bp)...\n",
            "[M::process] read 199352 sequences (20000086 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (17, 78456, 62, 8)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (205, 296, 514)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1132)\n",
            "[M::mem_pestat] mean and std.dev: (280.29, 119.67)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1441)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (261, 347, 472)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 894)\n",
            "[M::mem_pestat] mean and std.dev: (374.30, 159.75)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1105)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (105, 736, 1574)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4512)\n",
            "[M::mem_pestat] mean and std.dev: (950.49, 1022.54)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5981)\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_process_seqs] Processed 199314 reads in 23.562 CPU sec, 14.604 real sec\n",
            "[M::process] read 199386 sequences (20000116 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (19, 78447, 50, 20)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (198, 453, 880)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2244)\n",
            "[M::mem_pestat] mean and std.dev: (505.50, 402.69)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2926)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (262, 348, 474)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 898)\n",
            "[M::mem_pestat] mean and std.dev: (375.36, 161.04)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1110)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (125, 743, 1840)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5270)\n",
            "[M::mem_pestat] mean and std.dev: (1012.60, 1000.10)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6985)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (197, 342, 476)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1034)\n",
            "[M::mem_pestat] mean and std.dev: (313.67, 158.69)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1313)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199352 reads in 24.840 CPU sec, 15.776 real sec\n",
            "[M::process] read 199360 sequences (20000071 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (19, 78379, 39, 19)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (129, 192, 259)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 519)\n",
            "[M::mem_pestat] mean and std.dev: (161.13, 68.31)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 649)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (260, 345, 470)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 890)\n",
            "[M::mem_pestat] mean and std.dev: (372.71, 159.65)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1100)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (95, 406, 1433)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4109)\n",
            "[M::mem_pestat] mean and std.dev: (656.69, 643.28)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5447)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (125, 227, 322)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 716)\n",
            "[M::mem_pestat] mean and std.dev: (206.94, 100.64)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 913)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199386 reads in 24.012 CPU sec, 15.019 real sec\n",
            "[M::process] read 199374 sequences (20000071 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (19, 78623, 48, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (194, 360, 579)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1349)\n",
            "[M::mem_pestat] mean and std.dev: (343.56, 272.30)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1734)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (257, 340, 457)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 857)\n",
            "[M::mem_pestat] mean and std.dev: (363.88, 152.32)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1057)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (70, 214, 1396)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4048)\n",
            "[M::mem_pestat] mean and std.dev: (706.85, 837.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5374)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (168, 294, 446)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1002)\n",
            "[M::mem_pestat] mean and std.dev: (293.31, 153.04)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1280)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199360 reads in 24.502 CPU sec, 15.332 real sec\n",
            "[M::process] read 199332 sequences (20000007 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 78650, 52, 16)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (258, 341, 460)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 864)\n",
            "[M::mem_pestat] mean and std.dev: (366.03, 153.36)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1066)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (129, 522, 1399)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3939)\n",
            "[M::mem_pestat] mean and std.dev: (776.02, 832.57)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5209)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (174, 272, 353)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 711)\n",
            "[M::mem_pestat] mean and std.dev: (242.57, 102.04)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 890)\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199374 reads in 24.989 CPU sec, 16.314 real sec\n",
            "[M::process] read 199400 sequences (20000031 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 78512, 58, 15)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (169, 176, 311)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 595)\n",
            "[M::mem_pestat] mean and std.dev: (214.79, 90.92)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 737)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (259, 344, 466)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 880)\n",
            "[M::mem_pestat] mean and std.dev: (370.05, 156.94)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1087)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (78, 199, 908)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2568)\n",
            "[M::mem_pestat] mean and std.dev: (513.96, 634.24)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3398)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (257, 283, 1124)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2858)\n",
            "[M::mem_pestat] mean and std.dev: (412.62, 339.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3725)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199332 reads in 23.789 CPU sec, 15.062 real sec\n",
            "[M::process] read 199384 sequences (20000089 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (18, 78427, 44, 14)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 411, 450)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1052)\n",
            "[M::mem_pestat] mean and std.dev: (295.76, 160.55)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1353)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (255, 337, 454)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 852)\n",
            "[M::mem_pestat] mean and std.dev: (361.65, 151.55)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1051)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 452, 1545)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4347)\n",
            "[M::mem_pestat] mean and std.dev: (802.80, 833.92)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5748)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (213, 444, 5625)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 16449)\n",
            "[M::mem_pestat] mean and std.dev: (1947.86, 2595.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 21861)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199400 reads in 23.528 CPU sec, 14.585 real sec\n",
            "[M::process] read 199302 sequences (20000104 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 78646, 39, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 269, 526)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1290)\n",
            "[M::mem_pestat] mean and std.dev: (301.15, 301.75)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1672)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (253, 334, 447)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 835)\n",
            "[M::mem_pestat] mean and std.dev: (356.52, 146.94)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1029)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (132, 626, 1572)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4452)\n",
            "[M::mem_pestat] mean and std.dev: (819.13, 765.51)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5892)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (158, 255, 500)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1184)\n",
            "[M::mem_pestat] mean and std.dev: (320.00, 246.44)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1526)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199384 reads in 24.649 CPU sec, 15.539 real sec\n",
            "[M::process] read 199372 sequences (20000183 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (23, 78369, 59, 10)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (203, 334, 751)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1847)\n",
            "[M::mem_pestat] mean and std.dev: (383.32, 289.06)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2395)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (263, 351, 480)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 914)\n",
            "[M::mem_pestat] mean and std.dev: (379.63, 164.32)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1131)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (116, 223, 1598)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4562)\n",
            "[M::mem_pestat] mean and std.dev: (753.88, 884.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6044)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (235, 268, 484)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 982)\n",
            "[M::mem_pestat] mean and std.dev: (324.56, 205.32)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1231)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199302 reads in 26.052 CPU sec, 17.537 real sec\n",
            "[M::process] read 199392 sequences (20000116 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 78611, 39, 11)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (147, 255, 470)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1116)\n",
            "[M::mem_pestat] mean and std.dev: (256.62, 145.51)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1439)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (261, 347, 473)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 897)\n",
            "[M::mem_pestat] mean and std.dev: (374.10, 160.38)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1109)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (81, 449, 1757)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5109)\n",
            "[M::mem_pestat] mean and std.dev: (913.29, 1028.35)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6785)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (163, 240, 406)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 892)\n",
            "[M::mem_pestat] mean and std.dev: (308.09, 207.87)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1140)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199372 reads in 25.127 CPU sec, 16.116 real sec\n",
            "[M::process] read 199324 sequences (20000153 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (14, 78287, 50, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (144, 408, 621)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1575)\n",
            "[M::mem_pestat] mean and std.dev: (359.42, 287.66)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2052)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (258, 341, 461)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 867)\n",
            "[M::mem_pestat] mean and std.dev: (365.93, 154.58)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1070)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (101, 374, 1892)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5474)\n",
            "[M::mem_pestat] mean and std.dev: (795.54, 906.49)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7265)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (267, 327, 982)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2412)\n",
            "[M::mem_pestat] mean and std.dev: (475.58, 338.17)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3127)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199392 reads in 23.864 CPU sec, 14.688 real sec\n",
            "[M::process] read 199334 sequences (20000087 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20, 78269, 48, 18)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (195, 231, 580)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1350)\n",
            "[M::mem_pestat] mean and std.dev: (312.72, 277.39)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1735)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (260, 346, 471)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 893)\n",
            "[M::mem_pestat] mean and std.dev: (372.85, 159.19)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1104)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (103, 179, 1500)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4294)\n",
            "[M::mem_pestat] mean and std.dev: (677.27, 808.27)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5691)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (96, 342, 681)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1851)\n",
            "[M::mem_pestat] mean and std.dev: (367.35, 320.27)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2436)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199324 reads in 24.780 CPU sec, 15.448 real sec\n",
            "[M::process] read 199404 sequences (20000146 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (13, 78476, 56, 13)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (201, 241, 565)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1293)\n",
            "[M::mem_pestat] mean and std.dev: (384.23, 322.61)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1675)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (262, 348, 475)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 901)\n",
            "[M::mem_pestat] mean and std.dev: (375.57, 160.61)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1114)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (120, 485, 1676)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4788)\n",
            "[M::mem_pestat] mean and std.dev: (811.24, 844.07)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6344)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (264, 319, 500)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 972)\n",
            "[M::mem_pestat] mean and std.dev: (296.18, 132.91)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1208)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199334 reads in 23.035 CPU sec, 14.072 real sec\n",
            "[M::process] read 199388 sequences (20000059 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (22, 78357, 49, 11)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (136, 329, 353)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 787)\n",
            "[M::mem_pestat] mean and std.dev: (254.75, 140.36)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1004)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (258, 342, 465)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 879)\n",
            "[M::mem_pestat] mean and std.dev: (368.20, 156.64)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1086)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (122, 284, 922)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2522)\n",
            "[M::mem_pestat] mean and std.dev: (562.43, 613.94)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3322)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (300, 357, 610)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1230)\n",
            "[M::mem_pestat] mean and std.dev: (340.90, 196.82)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1540)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199404 reads in 23.413 CPU sec, 14.248 real sec\n",
            "[M::process] read 104194 sequences (10454520 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 78451, 50, 17)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (238, 323, 956)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2392)\n",
            "[M::mem_pestat] mean and std.dev: (547.55, 461.60)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3110)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (255, 338, 456)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 858)\n",
            "[M::mem_pestat] mean and std.dev: (362.59, 152.91)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1059)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (104, 245, 751)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2045)\n",
            "[M::mem_pestat] mean and std.dev: (454.49, 479.64)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2692)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (204, 420, 1231)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3285)\n",
            "[M::mem_pestat] mean and std.dev: (584.50, 446.09)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 4312)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 199388 reads in 22.245 CPU sec, 13.528 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 41059, 28, 10)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (257, 338, 455)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 851)\n",
            "[M::mem_pestat] mean and std.dev: (362.69, 150.60)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1049)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (111, 163, 574)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1500)\n",
            "[M::mem_pestat] mean and std.dev: (250.50, 243.42)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1963)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (154, 247, 444)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1024)\n",
            "[M::mem_pestat] mean and std.dev: (230.00, 114.74)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1314)\n",
            "[M::mem_pestat] skip orientation RF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 104194 reads in 11.389 CPU sec, 5.931 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem -t 2 /content/variant_calling/reference/reference.fasta /content/variant_calling/cleaned_data/SRR10527349_1.clean.fastq.gz /content/variant_calling/cleaned_data/SRR10527349_2.clean.fastq.gz\n",
            "[main] Real time: 234.815 sec; CPU: 374.424 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 6. SAM to BAM Conversion\n",
        "\n",
        "Convert SAM to BAM (binary format), sort, and index for efficient processing:\n",
        "\n",
        "**SAM to BAM Conversion Importance**\n",
        "\n",
        "1. **Storage Space**\n",
        "   - BAM is binary (compressed) format\n",
        "   - Takes 3-4x less space than SAM\n",
        "   - Crucial for large datasets\n",
        "\n",
        "2. **Processing Speed**\n",
        "   - BAM files process faster\n",
        "   - Binary format = quicker computer reading\n",
        "   - Essential for high-throughput analysis\n",
        "\n",
        "3. **Required for Tools**\n",
        "   - Most tools require BAM format\n",
        "   - Sorting and indexing needs BAM\n",
        "   - Standard in variant calling pipelines\n",
        "\n",
        "4. **Data Security**\n",
        "   - Less chance of corruption\n",
        "   - Maintains data integrity\n",
        "   - Checksums can verify content\n",
        "\n",
        "5. **Sorting & Indexing**\n",
        "   - BAM files can be sorted by position\n",
        "   - Enables creation of index files\n",
        "   - Allows quick access to specific regions\n",
        "\n",
        "6. **Compatibility**\n",
        "   - Industry standard format\n",
        "   - Works with all major tools\n",
        "   - Required for downstream analysis\n",
        "\n",
        "7. **Memory Efficiency**\n",
        "   - Better memory handling\n",
        "   - Faster data streaming\n",
        "   - Efficient parallel processing"
      ],
      "metadata": {
        "id": "feFiiKxo3CFC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Process Sample 1\n",
        "!samtools view -bS /content/variant_calling/alignments/SRR10527348.sam > /content/variant_calling/alignments/SRR10527348.bam\n",
        "!samtools sort /content/variant_calling/alignments/SRR10527348.bam -o /content/variant_calling/alignments/SRR10527348.sorted.bam\n",
        "!samtools index /content/variant_calling/alignments/SRR10527348.sorted.bam\n",
        "!rm /content/variant_calling/alignments/SRR10527348.sam /content/variant_calling/alignments/SRR10527348.bam"
      ],
      "metadata": {
        "id": "CATiPPHP3IKD",
        "outputId": "8f074280-1e68-43ed-8642-04cb3a4880c8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Process Sample 2\n",
        "!samtools view -bS /content/variant_calling/alignments/SRR10527349.sam > /content/variant_calling/alignments/SRR10527349.bam\n",
        "!samtools sort /content/variant_calling/alignments/SRR10527349.bam -o /content/variant_calling/alignments/SRR10527349.sorted.bam\n",
        "!samtools index /content/variant_calling/alignments/SRR10527349.sorted.bam\n",
        "!rm /content/variant_calling/alignments/SRR10527349.sam /content/variant_calling/alignments/SRR10527349.bam"
      ],
      "metadata": {
        "id": "PBQ06Eje3OaY",
        "outputId": "a0176326-40ab-4be4-a2be-07fdb8c13148",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 7. Variant Calling with BCFtools\n",
        "\n",
        "Call variants from the aligned reads.\n",
        "\n",
        "Input: Sorted BAM files + reference\n",
        "\n",
        "Output: VCF file (Variant Call Format)\n",
        "\n",
        "Example VCF format:\n",
        "CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO\n",
        "20      14370   rs6054257       G       A       29      PASS    NS=3;DP=14;AF=0.5\n",
        "\n",
        "**Variant Calling Overview**\n",
        "\n",
        "1. **Definition**\n",
        "   - Process of identifying DNA sequence variations\n",
        "   - Compares sample reads to reference genome\n",
        "   - Detects SNPs, insertions, deletions, etc.\n",
        "\n",
        "2. **Key Requirements**\n",
        "   - Quality-filtered reads\n",
        "   - Properly aligned BAM files\n",
        "   - Indexed reference genome\n",
        "   - Sufficient read depth/coverage\n",
        "\n",
        "3. **Basic Steps**\n",
        "   - Pile up reads at each position\n",
        "   - Calculate allele frequencies\n",
        "   - Apply quality filters\n",
        "   - Determine genotypes\n",
        "   - Estimate confidence scores\n",
        "\n",
        "4. **Common Variants Types**\n",
        "   - SNPs (Single Nucleotide Polymorphisms)\n",
        "   - Indels (Insertions/Deletions)\n",
        "   - Structural Variants\n",
        "   - Copy Number Variations\n",
        "\n",
        "5. **Output Format (VCF)**\n",
        "   - Chromosome\n",
        "   - Position\n",
        "   - Reference allele\n",
        "   - Alternative allele\n",
        "   - Quality scores\n",
        "   - Additional annotations\n",
        "\n",
        "6. **Important Metrics**\n",
        "   - Read depth (coverage)\n",
        "   - Mapping quality\n",
        "   - Base quality\n",
        "   - Genotype quality\n",
        "   - Allele frequency\n",
        "\n",
        "7. **Common Filters**\n",
        "   - Minimum read depth\n",
        "   - Minimum quality score\n",
        "   - Strand bias\n",
        "   - Mapping quality\n",
        "   - Allele balance"
      ],
      "metadata": {
        "id": "z6dwlVBB3Rpz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Prepare for variant calling\n",
        "%cd /content/variant_calling/variants\n",
        "\n",
        "# Create list of BAM files\n",
        "!ls /content/variant_calling/alignments/*.sorted.bam > bam_list.txt"
      ],
      "metadata": {
        "id": "9UARaFKc3dfk",
        "outputId": "484354ec-323e-42c4-ea08-7fea715a56ac",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/variant_calling/variants\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Run variant calling\n",
        "!bcftools mpileup \\\n",
        "    -f /content/variant_calling/reference/reference.fasta \\\n",
        "    -b bam_list.txt \\\n",
        "    -d 1000 \\\n",
        "    -q 20 \\\n",
        "    -Q 20 \\\n",
        "    --annotate FORMAT/AD,FORMAT/DP \\\n",
        "    -Ou | \\\n",
        "    bcftools call -mv --ploidy 2 -P 0.01 -Oz \\\n",
        "    -o variants.vcf.gz"
      ],
      "metadata": {
        "id": "ulxaa3WYCFfm",
        "outputId": "b8e4155b-6096-4378-85e1-c7b6e4427ca0",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 2 samples in 2 input files\n",
            "[mpileup] maximum number of reads per input file set to -d 1000\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Index VCF\n",
        "!bcftools index variants.vcf.gz"
      ],
      "metadata": {
        "id": "VbAkqtgyCHtc"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here are the key points about VCF filtering importance:\n",
        "\n",
        "- Removes low-quality variant calls to reduce false positives\n",
        "- Ensures adequate sequencing coverage by filtering based on read depth\n",
        "- Increases reliability of downstream genetic analyses\n",
        "- Makes data more manageable by reducing file size\n",
        "- Maintains high-confidence variants for research validity"
      ],
      "metadata": {
        "id": "RAxHOQeXHDA5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter variants\n",
        "!bcftools filter \\\n",
        "    -i 'QUAL>=20 && FORMAT/DP>=10' \\\n",
        "    -O z \\\n",
        "    -o variants.filtered.vcf.gz \\\n",
        "    variants.vcf.gz\n",
        "\n",
        "!bcftools index variants.filtered.vcf.gz"
      ],
      "metadata": {
        "id": "MbHA3BazCMF6"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 8. Examining output files to understand output file format\n",
        "\n",
        "1. fastp_report.html: Quality control metrics and graphs\n",
        "2. cleaned_R1/R2.fastq.gz: Filtered reads\n",
        "3. sorted.bam: Aligned reads in binary format\n",
        "4. variants.vcf: Final variants in VCF format\n",
        "\n",
        "Key VCF Fields:\n",
        "- CHROM: Chromosome\n",
        "- POS: Position in reference\n",
        "- REF: Reference base(s)\n",
        "- ALT: Alternative base(s)\n",
        "- QUAL: Quality score\n",
        "- FILTER: Whether variant passed filters\n"
      ],
      "metadata": {
        "id": "G5H1so3Z5mi6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Check alignment statistics\n",
        "print(\"Checking alignment statistics...\")\n",
        "!samtools flagstat /content/variant_calling/alignments/SRR10527348.sorted.bam\n",
        "!samtools flagstat /content/variant_calling/alignments/SRR10527349.sorted.bam"
      ],
      "metadata": {
        "id": "dfWMRbBV3n5J",
        "outputId": "c19c06e9-8c0b-4d1a-c772-fa7a20521381",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Checking alignment statistics...\n",
            "4345897 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "4335192 + 0 primary\n",
            "0 + 0 secondary\n",
            "10705 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "3780890 + 0 mapped (87.00% : N/A)\n",
            "3770185 + 0 primary mapped (86.97% : N/A)\n",
            "4335192 + 0 paired in sequencing\n",
            "2167596 + 0 read1\n",
            "2167596 + 0 read2\n",
            "3684122 + 0 properly paired (84.98% : N/A)\n",
            "3740144 + 0 with itself and mate mapped\n",
            "30041 + 0 singletons (0.69% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
            "3101667 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "3094612 + 0 primary\n",
            "0 + 0 secondary\n",
            "7055 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "2587222 + 0 mapped (83.41% : N/A)\n",
            "2580167 + 0 primary mapped (83.38% : N/A)\n",
            "3094612 + 0 paired in sequencing\n",
            "1547306 + 0 read1\n",
            "1547306 + 0 read2\n",
            "2515044 + 0 properly paired (81.27% : N/A)\n",
            "2555226 + 0 with itself and mate mapped\n",
            "24941 + 0 singletons (0.81% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check variant counts\n",
        "print(\"\\nChecking variant counts:\")\n",
        "!echo \"Raw variants:\"\n",
        "!bcftools view /content/variant_calling/variants/variants.vcf.gz | grep -v \"^#\" | wc -l\n",
        "!echo \"Filtered variants:\"\n",
        "!bcftools view /content/variant_calling/variants/variants.filtered.vcf.gz | grep -v \"^#\" | wc -l"
      ],
      "metadata": {
        "id": "88SLpF8XCZYz",
        "outputId": "4b216c3a-270a-46eb-b40b-7b8add604850",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Checking variant counts:\n",
            "Raw variants:\n",
            "55160\n",
            "Filtered variants:\n",
            "52848\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# View first few variants\n",
        "print(\"\\nFirst few variants:\")\n",
        "!bcftools view /content/variant_calling/variants/variants.filtered.vcf.gz | head -n 50"
      ],
      "metadata": {
        "id": "1FhojZt0Ce7A",
        "outputId": "a8e902d3-c2f4-4de3-8644-b403a00250ff",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "First few variants:\n",
            "##fileformat=VCFv4.2\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##bcftoolsVersion=1.13+htslib-1.13+ds\n",
            "##bcftoolsCommand=mpileup -f /content/variant_calling/reference/reference.fasta -b bam_list.txt -d 1000 -q 20 -Q 20 --annotate FORMAT/AD,FORMAT/DP -Ou\n",
            "##reference=file:///content/variant_calling/reference/reference.fasta\n",
            "##contig=<ID=NC_000913.3,length=4641652>\n",
            "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
            "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
            "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of raw reads supporting an indel\">\n",
            "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of raw reads supporting an indel\">\n",
            "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
            "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
            "##INFO=<ID=RPBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Read Position Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=BQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Base Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQSBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality vs Strand Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=SCBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Soft-Clip Length Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
            "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
            "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
            "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
            "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Number of high-quality bases\">\n",
            "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths (high-quality bases)\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
            "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
            "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
            "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Average mapping quality\">\n",
            "##bcftools_callVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_callCommand=call -mv --ploidy 2 -P 0.01 -Oz -o variants.vcf.gz; Date=Sun Feb  8 05:34:32 2026\n",
            "##bcftools_filterVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_filterCommand=filter -i 'QUAL>=20 && FORMAT/DP>=10' -O z -o variants.filtered.vcf.gz variants.vcf.gz; Date=Sun Feb  8 05:37:52 2026\n",
            "##bcftools_viewVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_viewCommand=view /content/variant_calling/variants/variants.filtered.vcf.gz; Date=Sun Feb  8 05:38:22 2026\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\t/content/variant_calling/alignments/SRR10527348.sorted.bam\t/content/variant_calling/alignments/SRR10527349.sorted.bam\n",
            "NC_000913.3\t91\t.\tG\tA\t492.639\tPASS\tDP=106;VDB=0.0892626;SGB=-1.37896;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,81,12;MQ=60\tGT:PL:DP:AD\t1/1:255,184,0:61:0,61\t1/1:255,96,0:32:0,32\n",
            "NC_000913.3\t393\t.\tT\tG\t492.639\tPASS\tDP=148;VDB=0.0521701;SGB=-1.38628;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,63,73;MQ=60\tGT:PL:DP:AD\t1/1:255,250,0:83:0,83\t1/1:255,160,0:53:0,53\n",
            "NC_000913.3\t588\t.\tG\tA\t492.639\tPASS\tDP=174;VDB=0.639346;SGB=-1.38629;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,69,91;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:91:0,91\t1/1:255,208,0:69:0,69\n",
            "NC_000913.3\t621\t.\tT\tC\t492.639\tPASS\tDP=182;VDB=0.00167441;SGB=-1.38629;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,48,123;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:99:0,99\t1/1:255,217,0:72:0,72\n",
            "NC_000913.3\t669\t.\tG\tA\t492.639\tPASS\tDP=196;VDB=0.999999;SGB=-1.38629;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,80,107;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:110:0,110\t1/1:255,232,0:77:0,77\n",
            "NC_000913.3\t690\t.\tC\tT\t492.639\tPASS\tDP=201;VDB=0.974389;SGB=-1.38629;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,93,96;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:120:0,120\t1/1:255,208,0:69:0,69\n",
            "NC_000913.3\t747\t.\tA\tG\t492.639\tPASS\tDP=194;VDB=0.999944;SGB=-1.37889;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,100,72;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:118:0,118\t1/1:255,163,0:54:0,54\n",
            "NC_000913.3\t759\t.\tC\tT\t492.639\tPASS\tDP=175;VDB=0.999981;SGB=-1.38585;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,95,66;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:106:0,106\t1/1:255,166,0:55:0,55\n",
            "NC_000913.3\t774\t.\tT\tC\t492.639\tPASS\tDP=175;VDB=0.99956;SGB=-1.3859;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,95,69;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:108:0,108\t1/1:255,169,0:56:0,56\n",
            "NC_000913.3\t798\t.\tG\tA\t492.639\tPASS\tDP=184;VDB=0.555958;SGB=-1.38628;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,81,83;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:103:0,103\t1/1:255,184,0:61:0,61\n",
            "NC_000913.3\t867\t.\tC\tT\t492.639\tPASS\tDP=183;VDB=0.683629;SGB=-1.38628;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,80,94;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:110:0,110\t1/1:255,193,0:64:0,64\n",
            "NC_000913.3\t870\t.\tC\tT\t492.639\tPASS\tDP=175;VDB=0.472056;SGB=-1.38627;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,77,94;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:109:0,109\t1/1:255,187,0:62:0,62\n",
            "NC_000913.3\t903\t.\tT\tC\t492.639\tPASS\tDP=191;VDB=0.970483;SGB=-1.38549;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,112,70;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:122:0,122\t1/1:255,181,0:60:0,60\n",
            "NC_000913.3\t939\t.\tG\tA\t492.639\tPASS\tDP=205;VDB=0.926446;SGB=-1.38625;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,102,90;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:125:0,125\t1/1:255,202,0:67:0,67\n",
            "NC_000913.3\t966\t.\tT\tC\t492.639\tPASS\tDP=173;VDB=0.775886;SGB=-1.38629;FS=0;MQ0F=0;AC=4;AN=4;DP4=0,0,89,71;MQ=60\tGT:PL:DP:AD\t1/1:255,255,0:98:0,98\t1/1:255,187,0:62:0,62\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Take home message\n",
        "Here are the key takeaways for the fastp to VCF generation pipeline:\n",
        "\n",
        "Input Files:\n",
        "- Raw FASTQ files (paired-end reads)\n",
        "- Reference genome FASTA file\n",
        "\n",
        "Key Steps:\n",
        "- Quality control with fastp (trimming adapters, low quality bases)\n",
        "- Read alignment to reference using BWA-MEM\n",
        "- Convert SAM to BAM format and sort\n",
        "- Variant calling with BCFtools to generate VCF\n",
        "- VCF filtering based on quality metrics\n",
        "\n",
        "Important Output Files:\n",
        "- Cleaned FASTQ files from fastp\n",
        "- QC reports showing read quality metrics\n",
        "- Sorted BAM files containing aligned reads\n",
        "- VCF file with filtered variants\n",
        "- Alignment statistics reports\n",
        "\n",
        "Best Practices:\n",
        "- Always check QC reports before proceeding\n",
        "- Index reference genome and BAM files\n",
        "- Sort BAM files before variant calling\n",
        "- Apply appropriate filters to raw VCF file\n",
        "- Keep track of intermediate files and reports"
      ],
      "metadata": {
        "id": "PhI9EUtq4cF2"
      }
    }
  ]
}